### ISSUE TITLE
Vagrant domain "waiting for an IP address" timeout (despite getting one)

### ISSUE THREAD

**Issue by peanutbutterandcrackers** [vagrant-libvirt/vagrant-libvirt#1211]
### Steps to reproduce
1. `vagrant up` # with the Vagrantfile reproduced below

### Expected behaviour
Normal VM bootup so one could move to `vagrant ssh`

### Actual behaviour
This issue might be somewhat related to #1097.

`vagrant up` gets stuck at "waiting for an IP address". But here's the thing: the VM is assigned an IP address, as confirmed by a look at the VM from virt-viewer, and an `nmap`. One can even `ping the.vm's.ip.address` and get the expected replies. However, `vagrant up` just gets stuck.

Here is something that I find quite suspicious, however. This is what the VM looks like in virtual machine manager (while it is stuck):
![Screenshot_appinventor-sources_default_2021-03-04_06:03:55](https://user-images.githubusercontent.com/26764547/109893347-9b371280-7cb3-11eb-8abb-822e96b6a721.png)
I wonder if it is an issue of the hostname service not being up on time. The nmap output also shows no hostname. And there are a few other issues too, it appears (as seen in the screenshot).

Few more things that might be worthy of note:
1. I'm trying to connect to a shared network device. A bridge.
2. I have ipv6 disabled on this bridge. But a previous attempt w/ ipv6 enabled, too, produced the same result.

### System configuration
**OS/Distro version:**: Linux Mint 20

**Libvirt version:** 5.8.0

**Output of `vagrant version; vagrant plugin list`:**
Installed Version: 2.2.6
vagrant-libvirt (0.0.45, system)

**Attach Output of `VAGRANT_LOG=debug vagrant ... --provider=libvirt >vagrant.log 2>&1`**
[vagrant.log](https://github.com/vagrant-libvirt/vagrant-libvirt/files/6080090/vagrant.log)

**A Vagrantfile to reproduce the issue:**
```
# -*- mode: ruby -*-
# vi: set ft=ruby :

# All Vagrant configuration is done below. The "2" in Vagrant.configure
# configures the configuration version (we support older styles for
# backwards compatibility). Please don't change it unless you know what
# you're doing.

Vagrant.configure("2") do |config|

  config.vm.box = "ubuntu/bionic64"

  config.vm.boot_timeout = 400

  config.vm.provider "virtualbox" do |v|
    v.name = "ForAppinventor2-bionic64"
    v.memory = "4096"
    v.customize ["modifyvm", :id, "--usb", "on"]
    # fix for slow network
    v.customize ["modifyvm", :id, "--natdnshostresolver1", "on"]
    v.customize ["modifyvm", :id, "--natdnsproxy1", "on"]
    v.customize ["modifyvm", :id, "--nictype1", "virtio"]
  end

  config.vm.provider :libvirt do |libvirt|
    config.vm.box = "generic/ubuntu1804"
    # By default, './' is synced onto '/vagrant' with NFS, which does not work for debian family images for some reason;
    # hence we use rsync. There are a few broken links in the source tree that will fail the default rsync syncing sche-
    # me, so we give explicit rsync args and omit "--copy-links".
    config.vm.synced_folder './', '/vagrant', type: 'rsync', rsync__args: ["--verbose", "--archive", "--update"]
    # # # # # # # # # # For QEMU/KVM user session # # # # # # # # # #
    # Reference:
    # https://github.com/vagrant-libvirt/vagrant-libvirt#qemu-session-support
    # Uncomment the following if connecting to QEMU/KVM user session:
    libvirt.qemu_use_session = true
    libvirt.uri = "qemu:///session"
    libvirt.system_uri = "qemu:///session"
    libvirt.management_network_device = 'foobar-bridge'
  end

  config.vm.provision :shell, path: "bootstrap.sh"

  config.vm.network :forwarded_port, guest: 8888, host: 8888
  config.vm.network :forwarded_port, guest: 9876, host: 9876
  config.vm.network :forwarded_port, guest: 9990, host: 9990
end

```

**Are you using upstream vagrant package or your distros package?**
Distro

**Comment by electrofelix** [comment]
I'm guessing that `foobar-bridge` is managed outside of libvirt and that address assigned to devices connected to it do not appear when the net-dhcp-leases is called. You can check what is shown by with by running `virsh net-dhcp-leases vagrant-default` when the VM is running.

The hostname will be set by vagrant once a usable ssh connection is available into the VM

**Comment by peanutbutterandcrackers** [comment]
@electrofelix - You are correct. `foobar-bridge` is managed outside of libvirt. Is there any work-around I could do to get this to work? I want to stick to this [work flow](https://xkcd.com/1172/). (:

**Comment by electrofelix** [comment]
Unfortunately not without a patch to how the ip_address is looked up, would possibly need to allow the user to provide a custom function that can replace the existing call. You can monkey patch the vagrant-libvirt code from your Vagrantfile, so it is possible to make it work the way you want.

**Comment by peanutbutterandcrackers** [comment]
Whoa. That seems to be out of my knowledge area. Could you please guide me to the required documentation and other references please? I'll spend a week or two reading everything up and will try that, and will get back to you with the results, if any.

As an aside, on the libvirt only side, however, I have been running able to use `foobar-bridge` just perfectly.

**Comment by electrofelix** [comment]
Yeah, the main problem is the need to identify the IP address assigned to the device in the guest so that vagrant can connect and execute and defined provisioners.

The underlying code involved is:
https://github.com/vagrant-libvirt/vagrant-libvirt/blob/bb2b022d3fb58375984f1d96c6a41c6481465b51/lib/vagrant-libvirt/driver.rb#L98-L105

Which in turn results in calls the following:
https://github.com/fog/fog-libvirt/blob/c927388af331e2760e0e7e6b39c41d1a63b2fc64/lib/fog/libvirt/models/compute/server.rb#L273-L299 that in turn uses the net-dhcp-leases.

The problem in this case is that it would be more desirable to be able to force execution of the ip_command as defined at https://github.com/vagrant-libvirt/vagrant-libvirt/blob/bb2b022d3fb58375984f1d96c6a41c6481465b51/lib/vagrant-libvirt/driver.rb#L38-L39

It's definitely enough work to try and establish a sensible order of preference here that I figure it'll be quite a while to be fixed, however as ruby doesn't close classes/modules fully after definition, it is possible to monkey patch them from the Vagrantfile.

I believe if you start your Vagrantfile with the following:
```
require 'vagrant-libvirt/driver'
module VagrantPlugins
  module ProviderLibvirt
    class Driver
      def get_ipaddress(machine)
        # you can now redefine what happens here and access the
        # mac address to search using the following
        domain = get_domain(@machine.id)
        mac = domain.mac

        # now perform the search for the correct IP address
      end
    end
  end
end

# Rest of Vagrantfile below
```
You'll be able to redefine the behaviour of the `get_ipaddress` function


Avoid trying to use `machine` that is passed into that function as I have a PR that will land at some stage that switches to using domain instead around here. Instead you'll be able to access the instance value for the current machine which should continue to work as expected.

I'll grant you this is a bit more exotic a bandaid than you were probably expecting, however it should provide a path to solve your immediate issue.

I suspect the full fix probably requires work in fog-libvirt to use the ip_command it it's explicitly provided and only use the net-dhcp-leases API otherwise with a fallback to using the ip command if the version of libvirt is too old. Alternatively the maintainers of that may feel that the default ip_command should be removed from fog-libvirt as something that is looking for the address outside of libvirt and we'll need to lift that code (including the part that works out whether to connect to the host via SSH or not) and maintain it in vagrant-libvirt as a method of supporting bridges and use of network devices not managed directly by libvirt.

Basically there's likely to be a bit of coordination required to land the necessary fixes, potentially it gets a bit easier if fog-libvirt are planning to remove the old support for ip_command as it can just move into this project, but then still a bit of work to determine the correct time to use it.

**Comment by peanutbutterandcrackers** [comment]
Whoa. Thank you very much for the guidance.

I'm afraid I still need to read up a bit before I can proceed. At the moment, I'm sort of caught up. But I will get back to this in a couple of days and will bother you with questions as they arise.

Thank you very much.

**Comment by electrofelix** [comment]
@peanutbutterandcrackers the recent 0.7.0 release contains support for the qemu agent, which is probably the correct solution for this problem.

Could you change your Vagrantfile definition to add the following to the libvirt provider config and see if it resolves the issue for you without needing to patch the existing vagrant-libvirt behaviour:
```
libvirt.qemu_use_agent = true
libvirt.channel :type => 'unix', :target_name => 'org.qemu.guest_agent.0', :target_type => 'virtio'
```

The next release will remove the need to specify `libvirt.channel :type => 'unix', :target_name => 'org.qemu.guest_agent.0', :target_type => 'virtio'` unless needing to control it specifically.
